homeassistant/components/google_assistant/trait.py

```diff
@@ -317,6 +332,67 @@ async def execute(self, command, data, params, challenge):
         """Execute a trait command."""
         raise NotImplementedError
 
+    def _translator(self, value: str | None, attribute: str) -> Callable[[str], str]:
+        entity_registry = er.async_get(self.hass)
+        domain = self.state.domain
+        device_class = self.state.attributes.get(ATTR_DEVICE_CLASS)
+        platform = None
+        translation_key = None
+        if entry := entity_registry.async_get(self.state.entity_id):
+            platform = entry.platform
+            translation_key = entry.translation_key
+
+        if attribute is not ATTR_STATE and value:
+            return partial(
+                async_translate_state_attribute_value,
+                self.hass,
+                attribute,
+                value,
+                domain,
+                platform,
+                translation_key,
+                device_class,
+            )
+
+        if attribute is not ATTR_STATE:
+            return partial(
+                async_translate_state_attribute_name,
+                self.hass,
+                attribute,
+                domain,
+                platform,
+                translation_key,
+                device_class,
+            )
+
+        if value:
+            return partial(
+                async_translate_state,
+                self.hass,
+                value,
+                domain,
+                platform,
+                translation_key,
+                device_class,
+            )
+
+        return partial(
+            async_translate_state_name,
+            self.hass,
+            domain,
+            platform,
+            translation_key,
+            device_class,
+        )
```

Comment:
This stuff ends up rather ugly. The way we need to look this stuff up is somewhat messy.